### PR TITLE
get problem summary into problem report

### PR DESCRIPTION
**Pull request by reinsch82**
Fixes https://github.com/gradle/gradle-private/issues/4539

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

File: platforms/ide/problems/src/main/java/org/gradle/problems/internal/services/DefaultProblemSummarizer.java [hunk 1]
```diff
@@ -62,17 +75,15 @@ public String getId() {
     @Override
     public void report(File reportDir, ProblemConsumer validationFailures) {
         List<Pair<ProblemId, Integer>> cutOffProblems = getCutOffProblems();
+        problemReportCreator.report(reportDir, cutOffProblems);
         eventEmitter.emitNow(currentBuildOperationRef.getId(), new DefaultProblemsSummaryProgressDetails(cutOffProblems));
     }
 
     private List<Pair<ProblemId, Integer>> getCutOffProblems() {
```
Comment (donat) [inline platforms/ide/problems/src/main/java/org/gradle/problems/internal/services/DefaultProblemSummarizer.java:82]: ❌ Please extract the Pair to a dedicated type

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ProblemReportCreator.java [hunk 2]
```diff
@@ -14,10 +14,17 @@
...
 
-public interface ProblemReportCreator extends ProblemReporter, ProblemEmitter {
+import java.io.File;
+import java.util.List;
+
+public interface ProblemReportCreator {
+
+    void report(File reportDir, List<Pair<ProblemId, Integer>> cutOffProblems);
+
+    void emit(Problem problem);
```
Comment (donat) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ProblemReportCreator.java:29]: ❌ rename method.

```
    /**
     * Stores a new problem in a temporary file that will be added to the final report when #createProblem is called.
    */
    void addProblem(Problem problem);

    /**
     * Renders a new problem report in the target directory. The report will contain the summaries of the omitted problem reports. The report file is only created if there's at least one problem reported.
    */
    void createReportFile(File reportDir, List<Pair<ProblemId, Integer>> problemSummaries);
```

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r812/ProblemThresholdCrossVersionTest.groovy [hunk 3]
```diff
@@ -53,7 +53,7 @@ class ProblemThresholdCrossVersionTest extends ToolingApiSpecification {
         then:
         def problems = listener.problems
         problems.size() == THRESHOLD_DEFAULT_VALUE
-        validateFirstNProblems(THRESHOLD_DEFAULT_VALUE, problems)
+        validateNToMProblems(0, THRESHOLD_DEFAULT_VALUE, problems)
```
Comment (donat) [inline platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r812/ProblemThresholdCrossVersionTest.groovy:56]: 🤔 
```suggestion
        validateProblemsRange(0, THRESHOLD_DEFAULT_VALUE, problems)
```

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemSummariesEvent.java [hunk 4]
```diff
@@ -39,6 +39,8 @@ public interface ProblemSummariesEvent extends ProblemEvent {
     /**
      * Get problems summaries.
      *
+     * There could more than one summary per problem id.
```
Comment (donat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemSummariesEvent.java:42]: ❌ 

```suggestion
     * There can be more than one summary for each problem id.
```

**Review by donat — CHANGES_REQUESTED** [review]
Looks good. I have a a few suggestions regarding naming and documentation.

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ProblemReportCreator.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+package org.gradle.api.problems.internal;
+
+import java.io.File;
+import java.util.List;
+
+public interface ProblemReportCreator {
+    /**
+     * Stores a new problem in a temporary file that will be added to the final report when #createProblem is called.
```
Comment (donat) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ProblemReportCreator.java:24]: Please make sure that that the text is valid  javadoc (I think the createProblem reference will not work).

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ProblemSummaryData.java [hunk 6]
```diff
@@ -14,10 +14,27 @@
...
+package org.gradle.api.problems.internal;
 
-import org.gradle.api.problems.internal.ProblemEmitter;
-import org.gradle.problems.buildtree.ProblemReporter;
+import org.gradle.api.problems.ProblemId;
 
-public interface ProblemReportCreator extends ProblemReporter, ProblemEmitter {
+/**
+ * Problem summary data
+ */
```
Comment (donat) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ProblemSummaryData.java:23]: Trivial comment is trivial. Does sanityCheck fail if you don't add this?
If so, there's a missing dot at the end of the sentence.

**Review by donat — APPROVED** [review]
Much more readable. I think it's ready for merge. I only have a 💄 comment.

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ProblemSummaryData.java (hunk 6 above)
Comment (reinsch82) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ProblemSummaryData.java:23]: That was part of me figuring out what is going on [here](https://gradle.slack.com/archives/CBSJ2GUTV/p1731947425228549). Trying to trigger the build again and getting a saner result.

**Pull request merged**